### PR TITLE
Consolidate Error Messages

### LINKED ISSUE #266: Evaluate message enums
A quick pass suggested that some of our message types are never used to instantiate a message, including
- `KeygenMessageType::PublicKeyShare`
- `AuxInfoMessageType::Public`

- [ ] Review whether these are needed
- [ ] Remove them if not

### PR DESCRIPTION
**Pull request by tyurek**
Addresses a component of #157 by further reducing the number of error types. The lost context is instead replaced by logging

Also closes #266

### CODE REVIEW

File: src/keygen/participant.rs [hunk 1]
```diff
@@ -122,7 +122,6 @@ impl ProtocolParticipant for KeygenParticipant {
             MessageType::Keygen(KeygenMessageType::R3Proof) => {
                 self.handle_round_three_msg(rng, message, input)
             }
-            MessageType::Keygen(_) => Err(InternalError::MessageMustBeBroadcasted)?,
```
Comment (tyurek) [inline src/keygen/participant.rs:125]: This appears to have been leftover code from before message types for broadcast changed

File: src/utils.rs [hunk 2]
```diff
@@ -207,6 +206,7 @@ pub(crate) fn random_bn_in_z_star<R: RngCore + CryptoRng>(
         .ok_or(RetryFailed)
 }
 
+// Returns x: BigNumber as a k256::Scalar mod k256_order
```
Comment (tyurek) [inline src/utils.rs:209]: The function already performed a modulo operation (I just made this clearer), so submitting too big of a BN shouldn't be an issue and any failures of this function should be bugs.

File: src/errors.rs [hunk 3]
```diff
@@ -23,7 +23,7 @@ pub enum InternalError {
     CallingApplicationMistake(#[from] CallerError),
     #[error("Serialization Error")]
     Serialization,
-    #[error("Protocol error")]
+    #[error("Some player sent a message which does not match the protocol specification")]
     ProtocolError,
     #[error("Could not successfully generate proof")]
     CouldNotGenerateProof,
```
Comment (tyurek) [inline src/errors.rs:29]: There's some work to do here in determining which (if any) of these are user error.
Also, the code that uses this error type makes ample use of the `warn! `macro to add context. If bugs here aren't user error, then these should be `error!` instead, however, making this change will result in relentless terminal spam upon running `cargo make ci`

File: src/errors.rs [hunk 4]
```diff
@@ -33,28 +33,14 @@ pub enum InternalError {
     InternalInvariantFailed,
     #[error("Paillier error: `{0}`")]
     PaillierError(#[from] paillier::Error),
```
Comment (tyurek) [inline src/errors.rs:35]: This one also involves a larger code rewrite, and may have similar issues to `CouldNotGenerateProof`

File: src/errors.rs [hunk 5]
```diff
@@ -23,7 +23,7 @@ pub enum InternalError {
     CallingApplicationMistake(#[from] CallerError),
     #[error("Serialization Error")]
     Serialization,
```
Comment (tyurek) [inline src/errors.rs:25]: I think this one is context dependent as to whether it's your fault or an external party's and so probably requires a bit of work

File: src/errors.rs [hunk 6]
```diff
@@ -33,28 +33,14 @@ pub enum InternalError {
     InternalInvariantFailed,
     #[error("Paillier error: `{0}`")]
     PaillierError(#[from] paillier::Error),
-    #[error("Failed to convert BigNumber to k256::Scalar, as BigNumber was not in [0,p)")]
-    CouldNotConvertToScalar,
-    #[error("Could not invert a Scalar")]
-    CouldNotInvertScalar,
     #[error("Reached the maximum allowed number of retries")]
     RetryFailed,
```
Comment (tyurek) [inline src/errors.rs:37]: This almost seems like it falls into a different category of error:  not enough resources or something similar? Though if the number of retries is hardcoded into the library, maybe this is just InternalInvariantFailed

File: src/errors.rs [hunk 7]
```diff
@@ -33,28 +33,14 @@ pub enum InternalError {
...
     #[error("This Participant was given a message intended for somebody else")]
     WrongMessageRecipient,
```
Comment (tyurek) [inline src/errors.rs:39]: User error or ProtocolError? Maybe it depends on how a Message is constructed from bytes over the wire?

File: src/errors.rs [hunk 8]
```diff
@@ -33,28 +33,14 @@ pub enum InternalError {
...
     #[error("Encountered a MessageType which was not expected in this context")]
     MisroutedMessage,
```
Comment (tyurek) [inline src/errors.rs:41]: Seems like this should be InternalInvariantFailed

File: src/errors.rs [hunk 9]
```diff
@@ -33,28 +33,14 @@ pub enum InternalError {
...
-    #[error("Could not construct signature from provided scalars")]
-    SignatureInstantiationError,
-    #[error("Tried to produce a signature without including shares")]
-    NoChainedShares,
     #[error("Storage does not contain the requested item")]
     StorageItemNotFound,
```
Comment (tyurek) [inline src/errors.rs:43]: Not really an error in the same way the others are. Maybe Storage should just return Options?

File: src/errors.rs [hunk 10]
```diff
@@ -33,28 +33,14 @@ pub enum InternalError {
...
-    #[error("The provided Broadcast Tag was not the expected tag for this context")]
-    IncorrectBroadcastMessageTag,
-    #[error("Encountered a Message sent directly, when it should have been broadcasted")]
-    MessageMustBeBroadcasted,
-    #[error("Broadcast has irrecoverably failed: `{0}`")]
-    BroadcastFailure(String),
     #[error(
         "Tried to start a new protocol instance with an Identifier used in an existing instance"
     )]
```
Comment (tyurek) [inline src/errors.rs:46]: User error or ProtocolError? Currently not used anywhere

File: src/auxinfo/proof.rs [hunk 11]
```diff
@@ -31,7 +31,7 @@ pub(crate) struct AuxInfoProof {
 impl AuxInfoProof {
     pub(crate) fn from_message(message: &Message) -> Result<Self> {
         if message.message_type() != MessageType::Auxinfo(AuxinfoMessageType::R3Proof) {
-            return Err(InternalError::IncorrectBroadcastMessageTag);
```
Comment (tyurek) [inline src/auxinfo/proof.rs:34]: I think this one just had the wrong error type

File: src/errors.rs (hunk 3 above)
Comment (tyurek) [reply src/errors.rs:29]: I almost want to say that ideally, at least for logging, there could be two types of InternalInvariantFailed: one which can be hit by test cases (such as by passing weird inputs to functions that shouldn't normally happen, like the current code that throws CouldNotGenerateProof errors), and one that should never be hit no matter what and will yell at you in angry red text if it is (like the current `error!` macro does).

So like four logging levels total:
info/debug: benign information about program flow
warn: not great but not necessarily catastrophic (detected an incorrect input from another player)
error: catastrophic, execution will not finish, something is wrong and it's probably our fault
ERROR: angry red text that should never appear, even in tests

But I'm not sure where to put "the were enough incorrect inputs from other parties that termination is no longer possible" in this taxonomy. Maybe each detected fault is a warn! and if the progam has to stop, it's an error! (generated at some higher level in the code)?

Thoughts?

Comment (tyurek) [reply src/errors.rs:29]: Actually, thinking about this some more, there shouldn't be that many ERRORs in practice and most InternalInvariantFaileds should be reachable by giving functions weird inputs. So maybe what we need is a way to mark a specific test as "this should fail" such that it doesn't print error! logs (but otherwise would, so that you see them in tests that shouldn't fail).

File: src/errors.rs (hunk 7 above)
Comment (marsella) [reply src/errors.rs:39]: Oops,I should have deleted this when I made the separate `CallerError` in #235. This shouldn't be used anywhere any more, but it's a caller error now because it's the responsibility of the calling application to route messages to the right participant.

File: src/errors.rs (hunk 9 above)
Comment (marsella) [reply src/errors.rs:41]: Now that all of our storage is of intermediate states and not of external values, this should be an `InternalInvariantFailed`. It should only happen if we fail to store stuff in our own protocol implementation.

File: src/errors.rs (hunk 10 above)
Comment (marsella) [reply src/errors.rs:44]: This can go away now that we have single-session Participants. The participant can no longer identify session-id-reuse errors.

File: src/keygen/participant.rs [hunk 12]
```diff
@@ -122,8 +122,10 @@ impl ProtocolParticipant for KeygenParticipant {
...
-            _ => Err(InternalError::MisroutedMessage)?,
+            _ => {
+                error!("Incorrect MessageType given to KeygenParticipant");
```
Comment (marsella) [inline src/keygen/participant.rs:126]: suggestion: log the message type received here

File: src/keygen/participant.rs (hunk 1 above)
Comment (marsella) [reply src/keygen/participant.rs:125]: Oh yeah, I made #266 because I thought there might be some extra conditions in here. Can you mark it as being closed by this PR?

File: src/presign/participant.rs [hunk 13]
```diff
@@ -212,7 +212,10 @@ impl ProtocolParticipant for PresignParticipant {
                 self.handle_round_three_msg(rng, message, input)
             }
 
-            _ => Err(InternalError::MisroutedMessage),
+            _ => {
+                error!("Incorrect MessageType given to PresignParticipant");
```
Comment (marsella) [inline src/presign/participant.rs:216]: suggestion: log the received message type here.

File: src/presign/record.rs [hunk 14]
```diff
@@ -65,7 +63,10 @@ impl TryFrom<RecordPair> for PresignRecord {
             return Err(InternalInvariantFailed);
         }
 
-        let delta_inv = Option::<Scalar>::from(delta.invert()).ok_or(CouldNotInvertScalar)?;
+        let delta_inv = Option::<Scalar>::from(delta.invert()).ok_or_else(|| {
+            error!("Could not invert delta as it is 0. Either you got profoundly unlucky or more likely there's a bug");
+            InternalInvariantFailed
+        })?;
```
Comment (marsella) [inline src/presign/record.rs:69]: question: some of those deltas that are summed up to make this are from other parties, right? I might say `ProtocolError` here.

File: src/protocol.rs [hunk 15]
```diff
@@ -195,7 +195,8 @@ impl SignatureShare {
             }
             (Some(prev_r), Some(new_r)) => {
                 if prev_r != new_r {
-                    return Err(InternalError::SignatureInstantiationError);
+                    error!("Failed to create signature as prev_r != new_r");
+                    return Err(InternalError::InternalInvariantFailed);
```
Comment (marsella) [inline src/protocol.rs:199]: suggestion: can we phrase this a bit more humanely? I think this is true:
> Failed to chain signature shares together because r-values were different. Got {prev_r}, expected {new_r}.

Also, I think this is a `ProtocolError`, since it's caused by a bad signature share from one of the parties.

File: src/protocol.rs [hunk 16]
```diff
@@ -217,10 +218,15 @@ impl SignatureShare {
...
-        k256::ecdsa::Signature::from_scalars(r, s)
-            .map_err(|_| InternalError::SignatureInstantiationError)
+        let r = self.r.ok_or_else(|| {
+            error!("Tried to produce a signature without including shares");
+            InternalError::InternalInvariantFailed
+        })?;
+
+        k256::ecdsa::Signature::from_scalars(r, s).map_err(|_| {
+            error!("Unable to create ECDSA signature from provided r and s");
+            InternalError::InternalInvariantFailed
```
Comment (marsella) [inline src/protocol.rs:228]: todo: I think both of these are `ProtocolError`s, since they're functions of data from other parties.

File: src/utils.rs (hunk 2 above)
Comment (marsella) [reply src/utils.rs:209]: Nice, thanks for clarifying.

**Review by marsella — CHANGES_REQUESTED** [review]
Thanks for taking this on! I have a few comments.

File: src/protocol.rs (hunk 16 above)
Comment (tyurek) [reply src/protocol.rs:228]: I think the first one is a result of forgetting to call `chain()` on your `SignatureShare`

Second one I think falls under the category "ultimately caused by bad inputs, but ones which should have been caught at an earlier check". What do you call that? I'm thinking `InternalInvariantFailed` is appropriate since it's really our oops if this gets hit, even with malicious inputs.

File: src/errors.rs (hunk 3 above)
Comment (marsella) [reply src/errors.rs:29]: Re: terminal spam. I thought that the tests only print stderr logs when they fail? For example, we don't see the seeds that every test prints out unless the test fails. Why does this logging behave differently? @gatoWololo 

I agree with your last comment that we should assume any IIF is reachable by some weird combination of circumstances (even if that involves messing with internal state or storage or something), and I'd rather not try to distinguish between "things we haven't figured out how to test" vs "things that should actually never happen". The latter case feels more like something we should design better abstractions around so it's not even an option in the code.

File: src/errors.rs (hunk 6 above)
Comment (marsella) [reply src/errors.rs:37]: I feel like this is a blameless error, but the user is the one who has to do something about it.

File: src/protocol.rs (hunk 16 above)
Comment (marsella) [reply src/protocol.rs:228]: Okay, that's fair. Let's do IIF then.

File: src/errors.rs (hunk 3 above)
Comment (gatoWololo) [reply src/errors.rs:29]: @marsella To give some context:
Cargo test attempts to intercept stdout and stderr. However, it is hard to do this across all OSs. So the implementation is rather poor and only catches straightforward things like calls to `println!` and `eprintln!`. Which is what our seed does:
```rust
eprintln!("seed: {seed:?}");
```
So this output is suppressed unless our test fails.

This doesn't work for our logging however, so the output leaks through and always shows up. Tom and I realized this a little bit ago, but didn't seem like a priority, (it took me a while to figure out this is what was happening!). So here we are now :face_exhaling: 

A couple of thoughts/opinions from me:
- For what its worth, I think the focus should be to have good error logging and reporting when actually using the library. And worry less about how logging behaves in Ci or tests.
- Currently, we enable logging on all tests by having every test set up logging at the very begging. An EZ:tm: solution would just be to turn off logging on our unit tests. The disadvantage here is that we would lose logging for tests (which only matters if a test fails and we want more info on "why?"). But if a test fails, and we really want the logging for further debugging, we could always edit that specific test initialize logging and re-run it. Which I think would be good enough? (Theoretically, we could end up with an irreproducible test, so we are not able to run it again with logging, but I don't think this edge case is worth worrying about).
- In the future we could consider writing logs from unit tests to a file, that way it doesn't spam the terminal and we could always check the file if needed, (for CI we would need to make sure the file is persisted after the run). This seemed low-priority (and unfun) to implement, so I haven't proposed it :sweat_smile: but "it shouldn't be too difficult..."

Comment (marsella) [reply src/errors.rs:29]: Ah, I didn't realize the catching of of stderr is flaky.

I think the EZ approach sounds perfect. It matches our current setup, where if a test fails you can go do some manual poking (fix the seed) to get more information. Right now the only source of randomness in the library is the RNG (we don't have any external parties or network flakiness or whatever to cause other types of unpredictability), so irreproducible tests are, I think, not a concern at this stage. When we do add such tests, we could consider a different logging setup for them (since I think they'd be part of e2e tests that would live in a separate directory from all our unit tests).

With this approach, we can switch all the error logs to be `error!` and modify `init_testing` to not log errors by default (and document / implement separately how to log them if you need to). Tom, it might be appropriate to put that into a separate issue along with making decisions about this whole error type (that is, we'd leave this as-is for now and address it in a separate issue).

Comment (gatoWololo) [reply src/errors.rs:29]: Great, this approach sounds good to me!

**Comment by tyurek** [comment]
I moved ProtocolAlreadyTerminated over to CallerError as this seems mainly like a message routing issue. I guess it could also be that some faulty party sent a duplicate message, but faulty parties could also be the cause of some of the other CallerErrors and you need to draw the line somewhere

**Comment by tyurek** [comment]
Two open questions:

I'm unsure how to resolve the StorageItemNotFound type, as I don't think it quite fits into InternalInvariantFailed as currently used.

Second is: I took a different approach for PaillierError and was looking for feedback. Essentially, I left the PaillierError types as is (in that it has its own Enum) but I made it separate from InternalError. The thinking behind this is that paillier.rs sort of acts as an independent module anyway and a lot of the functions can be mapped into InternalInvariantFailed or ProtocolError depending on if they are called within `prove()` or `verify()` functions. 

The drawback here is that there's now a bunch of `map_err()` calls everywhere

File: src/errors.rs (hunk 9 above)
Comment (marsella) [reply src/errors.rs:29]: Good catch. Here's what I think:

 There are two retrieval functions on `LocalStorage`: `retrieve` and `retrieve_mut`.
 - `retrieve` is called by other functions. A failure in `retrieve` is never conditioned on -- it's propagated up to the caller. From a first pass, it looks like these failures are caused by an internal control flow problem -- like, we called the `gen_round_X_msgs` function but the prerequisites were not actually in place. To me, that means it's an `InternalInvariantFailed`, but it's not guaranteed -- it depends on how it's called. I would add guidance to that function to say, if you need to check whether these things exist, use the `contains` or `contains_for_all_ids` function first, and switch to IIF because 
 - `retrieve_mut` is called by `get_from_storage` only. The caller always checks the error case and returns a default. I think this function should be rewritten to return an `Option` instead of creating and logging an error, since it's not actually an error the way it's used.

Comment (marsella) [reply src/errors.rs:29]: Since this involves a bit of a refactor, I made #310 with these steps in it.

File: src/keygen/participant.rs [hunk 17]
```diff
@@ -166,7 +166,9 @@ impl ProtocolParticipant for KeygenParticipant {
         info!("Processing keygen message.");
 
         if *self.status() == Status::TerminatedSuccessfully {
-            return Err(InternalError::ProtocolAlreadyTerminated);
+            return Err(InternalError::CallingApplicationMistake(
+                CallerError::ProtocolAlreadyTerminated,
+            ));
```
Comment (marsella) [inline src/keygen/participant.rs:171]: nit: simplify pattern with `?` as above.

File: src/presign/participant.rs [hunk 18]
```diff
@@ -241,7 +241,9 @@ impl ProtocolParticipant for PresignParticipant {
         info!("Processing presign message.");
 
         if *self.status() == Status::TerminatedSuccessfully {
-            return Err(InternalError::ProtocolAlreadyTerminated);
+            return Err(InternalError::CallingApplicationMistake(
+                CallerError::ProtocolAlreadyTerminated,
+            ));
```
Comment (marsella) [inline src/presign/participant.rs:246]: nit: simplify pattern with `?`

**Review by marsella — CHANGES_REQUESTED** [review]
Thanks for following up here. It's nice to see so many of those error variants disappear. 

My comments are a mix here: some are minor code nits to change. I created a few follow-up issues for tasks that I think are big enough, because I don't want this PR to keep blowing up or living forever. 

I see what you were going to with the change to `PaillierError` but aesthetically I don't really like it. The `map_err`s make the code much less readable to me. Did you take time to distinguish between internal vs external errors in there? I think it's a big task to do it carefully and correctly (e.g. did encryption fail because the input was out of range or because randomness failed? In the former case, was it an input we chose or one that another party chose?) -- I can try to audit it here but it'll take some time.

My instinct is to keep the current status quo of returning a (somewhat useless) `paillier::Error` directly to the caller, and refactor it in a separate PR where we can spend more time thinking carefully about that problem by itself. @gatowololo can you weigh in before Tom makes a bunch of changes though?

**Comment by gatoWololo** [comment]
> My instinct is to keep the current status quo of returning a (somewhat useless) paillier::Error directly to the caller, and refactor it in a separate PR where we can spend more time thinking carefully about that problem by itself. @gatoWololo can you weigh in before Tom makes a bunch of changes though?

I'm split on this one. (Maybe I'm used to the `.map_err`-style of error handling so I don't mind it so much). How much more useful is it to return an `InternalInvariantError` vs the underlying paillier error? Either way I don't think it is super critical to have perfect errors returned. As long as we are doing the right checks and returning some error that will probably be good enough on the calling application.

**Comment by marsella** [comment]
> > My instinct is to keep the current status quo of returning a (somewhat useless) paillier::Error directly to the caller, and refactor it in a separate PR where we can spend more time thinking carefully about that problem by itself. @gatoWololo can you weigh in before Tom makes a bunch of changes though?
> 
> I'm split on this one. (Maybe I'm used to the `.map_err`-style of error handling so I don't mind it so much). How much more useful is it to return an `InternalInvariantError` vs the underlying paillier error? Either way I don't think it is super critical to have perfect errors returned. As long as we are doing the right checks and returning some error that will probably be good enough on the calling application.

Okay, thanks. Let's keep the change Tom made, then, and I'll make a lower-priority issue to revisit this in the future to see if we can simplify the error types returned from Paillier.

File: src/auxinfo/auxinfo_commit.rs [hunk 19]
```diff
@@ -26,7 +26,12 @@ pub(crate) struct AuxInfoCommit {
...
     pub(crate) fn from_message(message: &Message) -> Result<Self> {
         if message.message_type() != MessageType::Auxinfo(AuxinfoMessageType::R1CommitHash) {
-            return Err(InternalError::MisroutedMessage);
+            error!(
+                "Encountered unexpected MessageType. Expected {:?}, Got {:?}",
+                MessageType::Auxinfo(AuxinfoMessageType::R1CommitHash),
+                message.message_type()
+            );
+            return Err(InternalError::InternalInvariantFailed);
         }
```
[the remark this answers is not in the document]
Comment (tyurek) [reply src/auxinfo/auxinfo_commit.rs:35]: Okay, so leave this for now then?

**Review by marsella — APPROVED** [review]
This looks good to me! Thanks for iterating on this, I think this issue was not well scoped, but it's an improvement for the external API, for sure.
